### PR TITLE
samples: net: echo_async: Add IPv6 support

### PR DESCRIPTION
**Pull request by pfalcon**
The main reason is allowing to run on 6LoWPAN devices, though also to
demonstrate IPv6 support with BSD Sockets in general.

Tested on 96b_carbon.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

### CODE REVIEW

**Comment by pfalcon** [comment]
Depends on https://github.com/zephyrproject-rtos/zephyr/pull/1184

**Comment by pfalcon** [comment]
Successful building for 96b_carbon depends on https://github.com/zephyrproject-rtos/zephyr/pull/1169

File: samples/net/sockets/echo/src/socket_echo.c [hunk 1]
```diff
@@ -24,24 +24,31 @@
...
+	};
 
-	serv = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+	serv = socket(AF_INET6, SOCK_STREAM, IPPROTO_TCP);
 
-	bind_addr.sin_family = AF_INET;
-	bind_addr.sin_addr.s_addr = htonl(INADDR_ANY);
-	bind_addr.sin_port = htons(4242);
-	bind(serv, (struct sockaddr *)&bind_addr, sizeof(bind_addr));
+//	bind_addr.sin6_family = AF_INET6;
```
Comment (jukkar) [inline samples/net/sockets/echo/src/socket_echo.c]: I think you should remove these comments.

File: samples/net/sockets/echo_async/src/socket_echo.c [hunk 2]
```diff
@@ -70,36 +76,72 @@ void pollfds_del(int fd)
...
 			}
 			int fd = pollfds[i].fd;
-			if (fd == serv) {
-				int client = accept(serv, (struct sockaddr *)&client_addr,
+			if (i < 2) {
+				/* If server socket */
+				int client = accept(fd, (struct sockaddr *)&client_addr,
 						    &client_addr_len);
-				inet_ntop(client_addr.sin_family, &client_addr.sin_addr,
+				void *addr = &((struct sockaddr_in *)&client_addr)->sin_addr;
```
Comment (jukkar) [inline samples/net/sockets/echo_async/src/socket_echo.c]: Declare variables in start of the block

**Review by jukkar — CHANGES_REQUESTED** [review]

File: samples/net/sockets/echo/src/socket_echo.c (hunk 1 above)
Comment (pfalcon) [inline samples/net/sockets/echo/src/socket_echo.c]: Oops, sorry, that's a stray local change. I didn't mean to change samples/net/sockets/echo (since there's no suitable way to make it work with both IPv4 and IPv6), only the echo_async sample. Will revert this change.

File: samples/net/sockets/echo_async/src/socket_echo.c (hunk 2 above)
Comment (pfalcon) [inline samples/net/sockets/echo_async/src/socket_echo.c]: Well, it is - declared at the start of block together with "int client" before it, and both are initialized at the same time. Adding blank line.

**Comment by pfalcon** [comment]
Pushed new version, removing stray changes which shouldn't be part of the release.

File: samples/net/sockets/echo_async/src/socket_echo.c (hunk 2 above)
Comment (jukkar) [inline samples/net/sockets/echo_async/src/socket_echo.c]: Sorry, my bad, I misread the diff and missed the removal of inet_ntop.

**Review by jukkar — DISMISSED** [review]

**Review by mbolivar — COMMENTED** [review]
LGTM, not tested.

**Comment by pfalcon** [comment]
With https://github.com/zephyrproject-rtos/zephyr/pull/1184 merged, this now passes CI. @jukkar , please reinstate your approval when you get a chance.

File: samples/net/sockets/echo_async/README.rst [hunk 3]
```diff
@@ -59,7 +59,8 @@ Running application on POSIX Host
 
 The same application source code can be built for a POSIX system, e.g.
 Linux. (Note: if you look at the source, you will see that the code is
-the same except the header files are different for Zephyr vs POSIX.)
+the same except the header files are different for Zephyr vs POSIX, and
+there're additional option to set for Linux to make a socket IPv6-only).
```
Comment (dbkinder) [inline samples/net/sockets/echo_async/README.rst]: there's an additional...

File: samples/net/sockets/echo_async/README.rst [hunk 4]
```diff
@@ -78,6 +79,7 @@ To test:
 .. code-block:: console
 
     $ telnet 127.0.0.1 4242
+    $ telnet ::1 4242
```
Comment (dbkinder) [inline samples/net/sockets/echo_async/README.rst]: Only one of these telnet connections is going to work, right? (IPV4 vs 6)   You might want to write something like this instead as:

     $ telnet 127.0.0.1 4242     # use this for IPV4 
     $ telnet ::1 4242           # or this for IPV6

File: samples/net/sockets/echo_async/README.rst [hunk 5]
```diff
@@ -21,7 +20,7 @@ Requirements
 ************
 
 - :ref:`networking_with_qemu`
-- or, a board with hardware networking
+- or, a board with hardware networking (including 6LoWPAN)
```
Comment (dbkinder) [inline samples/net/sockets/echo_async/README.rst]: I could read this as you needed a board with 6LoWPAN hardware networking but I think you mean this:

     - or, a board with hardware networking (such as 6LoWPAN)

**Review by dbkinder — CHANGES_REQUESTED** [review]
suggested edits

File: samples/net/sockets/echo_async/README.rst (hunk 3 above)
Comment (pfalcon) [inline samples/net/sockets/echo_async/README.rst]: Fixed.

That option is actually Linux-specific, not part of POSIX. At the beginning of this paragraph it says "a POSIX system, e.g. Linux", so hopefully even a reader not closely familiar with "family tree" of the OSes won't be too confused how it all relates together. (And for someone familiar, the above gives "absolutely true" picture without simplifications.)

File: samples/net/sockets/echo_async/README.rst (hunk 4 above)
Comment (pfalcon) [inline samples/net/sockets/echo_async/README.rst]: Yes, they're intended to be run sequentially (with extra interaction for each), or in different terminal windows. Added comments (for similar commands earlier in the doc too).

File: samples/net/sockets/echo_async/README.rst (hunk 5 above)
Comment (pfalcon) [inline samples/net/sockets/echo_async/README.rst]: By adding text in parens I meant that this sample was actually tested on 6LoWPAN hardware. (Which is by definition IPv6 only, not supporting IPv4. Before I tested it, I wasn't sure if the sample would work out of the box in such case, and after I tested it, I wanted to give users a hint that it's supported, so for example, if it doesn't, it's not the sample limitation, but some other issue. (Note that another sample mentioned in the doc, sockets/echo, doesn't run on 6LoWPAN)).

**Review by dbkinder — APPROVED** [review]
Thanks for clarifying

**Comment by pfalcon** [comment]
Incorporated explicit fixes, and commented on other cases - I'm trying to give maximum info without overloading a reader, so folks who need specific points (e.g. whether the sample was tested on 6lowpan) will find their hints, and others can just skip thru them.

**Review by jukkar — APPROVED** [review]
